langchain4j-agentic/src/main/java/dev/langchain4j/agentic/internal/AgentInvoker.java

```
@@ -14,15 +19,32 @@
 
 public interface AgentInvoker extends AgentSpecification {
 
+    Logger LOG = LoggerFactory.getLogger(AgentInvoker.class);
+
     Method method();
 
     String toCard();
 
-    Object[] toInvocationArguments(AgenticScope agenticScope) throws MissingArgumentException;
+    AgentInvocationArguments toInvocationArguments(AgenticScope agenticScope) throws MissingArgumentException;
+
+    default Object invoke(AgenticScope agenticScope, Object agent, AgentInvocationArguments args) throws AgentInvocationException {
+        try {
+            beforeInvocation(new AgentRequest(agenticScope, name(), args.namedArgs()));
+        } catch (Exception e) {
+            LOG.error("Before agent invocation listener for agent " + name() + " failed: " + e.getMessage(), e);
+        }
+        Object result = internalInvoke(agent, args);
+        try {
+            afterInvocation(new AgentResponse(agenticScope, name(), args.namedArgs(), result));
+        } catch (Exception e) {
+            LOG.error("After agent invocation listener for agent " + name() + " failed: " + e.getMessage(), e);
+        }
```

I don't think you want to enable a silent failure here if the client throws an exception (whether deliberately or otherwise). This isn't the behavior I'd expect as a client and think it's generally bad practice for the system to swallow exceptions like this since it hides errors. Why not just let the system fail?

---

I discussed this with @dliubarskyi and he clarified here https://github.com/langchain4j/langchain4j/pull/3718#discussion_r2368683694 the rationals behind this. Basically we believe that an observability issue shouldn't be enough to break the main flow of the application.

---

That feels like it should be the client's decision since systems have different standards around error handling and not everyone will use this only for side effect free observability. Catching the exception this way basically takes away the clients ability to let unexpected exceptions propagate, if that's their standard. If someone doesn't want to crash and instead wants to log they can easily still do that through a simple try catch in the callback. 

In our system even if the component responsible only for streaming a status update produces an unexpected exception I would not tolerate that being silently logged, if we had control over the code.